crates/client/db/src/rocksdb_options.rs

```diff
@@ -1,3 +1,50 @@
+//! # Rocksdb options
+//!
+//! We configure the rocksdb database in a very specific way. We want to guarantee a few things:
+//!
+//! ## Fault tolerance
+//!
+//! The madara node can be stopped at any time, and it should be able to recover when restarting.
+//! In particular, in case of a power/hardware failure, the database should not be corrupted beyond repair. We cannot
+//! run code that would allow us to gracefully shutdown in these cases.
+//!
+//! The way rocksdb ensures this is usually by using a WAL (write-ahead log) and doing a 2PC (two-phase commit) under
+//! the hood. There is however another way to achieve fault tolerance with rocksdb, which is by enabling the [atomic flush]
+//! option.
+//!
+//! ## Consistent view of the database
+//!
+//! The usual way it is achieved is by batching all of the changes into a single batch or a single rocksdb transaction so
+//! that they are applied to the database atomically.
+//!
+//! ## Multithreaded commit to the database
+//!
+//! This is main reason we differ from the default rocksdb configuration: rocksdb Transaction objects are not thread safe,
+//! and committing a transaction also isn't multithreaded at all. This means that if we want to write a whole block into the
```

Comment:
> also isn't multithreaded at all

You mean, "isn't also thread-safe"?

Reply:
i actually mean the commit is done in a single-threaded fashion

Well, in a write-prepared transaction (the default), all of the changes are buffered in memory and then calling commit() on the transaction will apply them using the current thread only. this is what I mean: the commit is not multithreaded (which makes sense - it would be weird for rocksdb to start using all of the threads when doing commit() on a transaction)
In a write-unprepared transaction, all of the changes are applied directly to the database instead of being buffered. The commit will just update the latest sequence number of the database
This is very close to what we need actually! but we're still hitting a problem, which is that the transaction object is not thread-safe (here I'm talking about the struct being `Sync`) - also it uses &mut self for all of its operations. That might be a problem of the rust `rocksdb` crate though, but looking at the official documentation https://github.com/facebook/rocksdb/wiki/WriteUnprepared-Transactions it doesn't look like they say it's thread safe on the C++ side as well. There's this "Key Locking" section on that page that mention threads and locking but I'm not sure I 100% understand if it's relevant or not. I haven't looked at that part of the rocksdb C++ code though.

Hope that clears up this section, rocksdb is sometimes a bit confusing. I'll try to be a bit more clear about that, thank you!